protocol/messenger_saved_address.go

```diff
@@ -0,0 +1,112 @@
+)
+
+func (m *Messenger) AddSavedAddress(ctx context.Context, sa wallet.SavedAddress) error {
+	updatedClock, err := m.savedAddressesManager.AddOrUpdateSavedAddress(sa)
+	if err != nil {
+		return err
+	}
+	return m.SyncNewSavedAddress(ctx, &sa, updatedClock)
+}
+
+func (m *Messenger) DeleteSavedAddress(ctx context.Context, chainID uint64, address gethcommon.Address) error {
+	updatedClock, err := m.savedAddressesManager.DeleteSavedAddress(chainID, address)
+	if err != nil {
+		return err
+	}
+	return m.SyncDeletedSavedAddress(ctx, chainID, address, updatedClock)
+}
+
+func (m *Messenger) garbageCollectRemovedSavedAddresses() error {
+	return m.savedAddressesManager.DeleteSoftRemovedSavedAddresses(time.Now().AddDate(0, 0, -30).Unix())
+}
+
+func (m *Messenger) dispatchSyncSavedAddress(ctx context.Context, syncMessage protobuf.SyncSavedAddress) error {
+	if !m.hasPairedDevices() {
+		return nil
+	}
+
+	clock, chat := m.getLastClockWithRelatedChat()
+	syncMessage.SyncClock = clock
+
+	encodedMessage, err := proto.Marshal(&syncMessage)
+	if err != nil {
+		return err
+	}
+
+	_, err = m.dispatchMessage(ctx, common.RawMessage{
+		LocalChatID:         chat.ID,
+		Payload:             encodedMessage,
+		MessageType:         protobuf.ApplicationMetadataMessage_SYNC_SAVED_ADDRESS,
+		ResendAutomatically: true,
+	})
+	if err != nil {
+		return err
+	}
+
+	chat.LastClockValue = clock
+	return m.saveChat(chat)
+}
+
+func (m *Messenger) SyncNewSavedAddress(ctx context.Context, savedAddress *wallet.SavedAddress, updateClock int64) error {
+	return m.dispatchSyncSavedAddress(ctx, protobuf.SyncSavedAddress{
+		Address:     savedAddress.Address.Bytes(),
+		Name:        savedAddress.Name,
+		Favourite:   savedAddress.Favourite,
+		ChainId:     savedAddress.ChainID,
+		UpdateClock: updateClock,
+	})
+}
+
+func (m *Messenger) SyncDeletedSavedAddress(ctx context.Context, chainID uint64, address gethcommon.Address, updateClock int64) error {
+	return m.dispatchSyncSavedAddress(ctx, protobuf.SyncSavedAddress{
+		Address:     address.Bytes(),
+		ChainId:     chainID,
+		UpdateClock: updateClock,
+		Removed:     true,
+	})
+}
+
+func (m *Messenger) syncSavedAddress(ctx context.Context, savedAddress wallet.SavedAddress, metadata wallet.SavedAddressMeta) (err error) {
+	if metadata.Removed {
+		if err = m.SyncDeletedSavedAddress(ctx, savedAddress.ChainID, savedAddress.Address, metadata.UpdateClock.Int64); err != nil {
+			return err
+		}
+	} else {
+		if err = m.SyncNewSavedAddress(ctx, &savedAddress, metadata.UpdateClock.Int64); err != nil {
+			return err
+		}
+	}
+	return
+}
+
+func (m *Messenger) handleSyncSavedAddress(state *ReceivedMessageState, syncMessage protobuf.SyncSavedAddress) (err error) {
+	address := gethcommon.BytesToAddress(syncMessage.Address)
+	if syncMessage.Removed {
+		_, err = m.savedAddressesManager.DeleteSavedAddressIfNewerUpdate(syncMessage.ChainId,
+			address, int64(syncMessage.SyncClock), syncMessage.UpdateClock)
```

Comment:
I would not cast to `int64`, it's probably ok, but best to keep uint64 and compare uint64, any reason to use `int64`? (the only one I can think of is the fact that we have `nullint64` and not `nullUint64` , but if that's the case, you should convert as close to the db as possible, no earlier)

Reply:
> ... any reason to use `int64`? (the only one I can think of is the fact that we have `nullint64` and not `nullUint64`

Exactly

> but if that's the case, you should convert as close to the db as possible, no earlier)

I was thinking the opposite, better to have the limitations exposed higher the call chain, closer to requirements, where we have enough information to know if we will be affected instead of hiding it deeper where we miss domain knowledge.

Reply:
I think it's best to keep uint64 everywhere, and we should not be using nullInt sql in the metadata struct, as that's more of a database concern, in the rest of the codebase I think for the most part with use the `sql.Null` just to pull from the db in temporary variable, but then we use native go types for our structs, there's a few examples I believe in the various persistence.go files

Another note is that the approach you used leads to some subtle bugs, for example in some code in the PR, when you convert a uint64 to a int64 and use them, you need to be careful about the value being wrapped around:

```
 dbUpdateClock <= updateClock
```

If both wraps around, it works correctly, but if only ones does, it would not be what you expect

https://go.dev/play/p/V49B-mtjcJ1

So we should honor the original type before doing any operation, and the easiest and less prone to error way is to only use nullint64 to pull/store from the db, but immediately convert back to the original type in my opinion

Reply:
In the end, me relying on the client to mind the limitations and not reach overflowing and wrapping around looks like a more error-prone solution. Thanks